enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,754 @@
+Errors in conversion will be reported on the Machine API resource in the `status.conditions`.
+
+The new webhook introduced by this enhancement will have `FailPolicy=Ignore`, and therefore should not impact cluster operations when failing.
+
+When the controller and webhook are restored, they should, unless there are persistent errors in need of human intervention, continue reconciling and restore the desired state.
+
+## Implementation History
+
+- [ ] Enhancement merged (YYYY-MM-DD)
+
+## Alternatives
+
+### Use an aggregated API server to handle Machine API and Cluster API
+
+As an alterative to synchronising resources post update, we could use an aggregated API server to implement both Machine API and Cluster API endpoints going forward.
+
+This aggregated API server would be responsible for handling all API calls for both API groups and would interact directly with etcd for storage of the resources.
+
+In this case, it could ensure conversion is synchronised and written to etcd atomically, in a single call.
+This would prevent any drift between resources and mean a number of the rules above would no longer be required.
+
+This alternative has been dismissed for a number of reasons:
+- Aggregated API servers are generally complex and introduce new failure modes for the entire cluster
+- We are not certain that it is possible to take over the API endpoints of an existing installed CRD
+- Interacting directly at the storage layer is complex and adds other risks that we do not want to accept
+- Handling errors writing multiple objects to etcd for a single API call becomes complex
+- We still need a way to decide which operator is authoritative
+
+### Machine API controllers write out Cluster API resources
+
+In this alternative, rather than converting between resources in a separate controller, the implementation of each Machine controller for Machine API would be rewritten.
+Instead of interacting directly with the cloud provider, the rewrite would convert the resources from Machine API to Cluster API and then write these to the API server.
+
+The controllers would become shims, leveraging Cluster API as a backend to implement their existing functionality.
+
+This alternative has been dismissed based on the following reasons:
+- This doesn't help our maintenance burden, we still have Machine API controllers to manage
+- It is unclear how to handle a lack of feature parity between the Machine API and Cluster API resources for this design
+- We prefer to keep the conversion logic in a centralised place so that we can run 1 controller in the future, instead of many
+
+### Machine API controllers convert to Cluster API internally
+
+In this solution, the Machine API controllers would be rewritten to convert Machine API resources internally to Cluster API resources.
+Once the conversion has been completed, the controllers would feed the converted Cluster API resources into the Cluster API methods, using them as a library.
+
```

Comment:
Although we currently do this in MAPO, it is highly provider-specific and not at all suitable as a general solution. As you point out, it also does nothing towards API conversion.

You could possibly remove this section, tbh.

Reply:
I included it because it had been suggested as an alternative in the past, I don't see much harm in it being here